File: algorithms/python/238_productarrayexecptself.py
from typing import List
import math

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        left = [1]
        right = [1]
        results = []
        
        n = len(nums)
        
        for i in range(n):
            left.append(left[i]*nums[i])
            right.append(right[i]*nums[n-i-1])
        
        for i in range(n):
            results.append(left[i]*right[n-i-1])
        
        return results

# Alternate approach using math.prod -> Time limit exceeded
# class Solution:
#     def productExceptSelf(self, nums: List[int]) -> List[int]:
#         results = []
        
#         for i in range(len(nums)):
#             result = math.prod(nums[:i]) * math.prod(nums[i+1:])
#             results.append(result)
        
#         return results
    
    
def main():
    sol = Solution()
    nums = [1, 2, 3, 4]
    print(sol.productExceptSelf(nums))
    
if __name__ == "__main__":
    main()
            

File: algorithms/Array_and_Hashing/cpp/238_productarrayexecptself.cpp
#include <vector>
#include <iostream>

using namespace std;

class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) 
    {
        // left : save left of nums[i]
        // right : save righ of num[i]
        vector<int> left;
        vector<int> right;

        // Sure product != 0 in begin
        left.push_back(1);
        right.push_back(1);

        int size = nums.size();

        for(int i = 0; i < size - 1; i++)
        {   
            // cout<<"i = "<<i<<endl;
            // cout<<"nums = "<<nums[i]<<endl;
            
            left.push_back(left[i] * nums[i]);
            right.push_back(right[i] * nums[size - i - 1]);
            // cout<<"left: "<<left[i + 1]<<" right :"<<right[i + 1]<<endl;
        }

        vector<int> results;
        for(int i = 0; i < size; i++)
        {   
            // results[i] = left[i] * right[i]
            results.push_back(left[i] * right[size - i - 1]);
        }

        return results;
    }
};

int main()
{
    Solution sol;
    vector<int> nums = {-1,1,0,-3,3};
    cout<<"Results: ";
    vector<int> results = sol.productExceptSelf(nums);
    for(int &re : results)
    {
        cout<<re<<" ";
    }
    return 0;
}